### PR TITLE
Event assert

### PR DESCRIPTION
**Pull request by cachedout**
This is a new mixin for the salt test suite that allows a developer to fire events onto the event bus and then assert that they were received.

The theory here is that this should allow for a framework to better test systems like reactors, beacons and engines which are very much event based. For example, one could use this to test a beacon which watches a file by enabling the beacon, changing the file and asserting that the beacon fired an appropriate event.

I'm sure this could be improved in a number of areas. There are some potential race conditions here, most notably between the event socket connecting and the test firing an event to be tested and a second condition between the forked proc gathering the event and pushing it onto the queue of the test runner which does the actual assertion. So far in my testing this has been somewhat resilient but improvements are welcome!

### CODE REVIEW

File: tests/integration/__init__.py [hunk 1]
```diff
@@ -1082,6 +1082,63 @@ def master_opts(self):
...
+            try:
+                events = event.get_event(full=False)
+            except Exception:
+                # This is broad but we'll see all kinds of issues right now
+                # if we drop the proc out from under the socket while we're reading
+                pass
+            q.put(events)
+
+    def assertMinionEventFired(self, tag):
+        #TODO
```
Comment (rallytime) [inline tests/integration/__init__.py]: Did you mean to leave this here?

Comment (cachedout) [inline tests/integration/__init__.py]: Yes.

**Comment by rallytime** [comment]
Neat! I think it would be really great to get some documentation written about the existence of this new mixin as well, maybe on this page?

https://docs.saltstack.com/en/latest/topics/development/tests/integration.html

That way people know they can write tests like this.

File: tests/integration/reactor/reactor.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        '''
+        Fire an event on the master and ensure
+        that it pings the minion
+        '''
+        # Create event bus connection
+        e = salt.utils.event.get_event('minion', sock_dir=self.minion_opts['sock_dir'], opts=self.minion_opts)
+
+        e.fire_event({'a': 'b'}, '/test_event')
+
+        self.assertMinionEventReceived({'a': 'b'})
```
Comment (rallytime) [inline tests/integration/reactor/reactor.py]: This new test isn't running/getting picked up by the test runner on Jenkins. You need to add the following to bottom of this file:
```
if __name__ == '__main__':
    from integration import run_tests
    run_tests(ReactorTest)
```

Comment (cachedout) [inline tests/integration/reactor/reactor.py]: It doesn't look like that did the trick to me. I think there's more too getting this to work but I'm trying to remember what it is. :]

Comment (rallytime) [inline tests/integration/reactor/reactor.py]: Oh! You need to add some stuff to the integration test runner to tell it to look for tests in this directory. There's some docs here: https://docs.saltstack.com/en/latest/topics/development/tests/integration.html#adding-new-directories

File: tests/runtests.py [hunk 3]
```diff
@@ -247,6 +254,7 @@ def validate_options(self):
                 self.options.wheel,
                 self.options.api,
                 self.options.minion,
+                self.optsions.reactor,
```
Comment (rallytime) [inline tests/runtests.py]: This should be `options`

**Comment by rallytime** [comment]
One spellcheck fix and a rebase, and we can get this in! (I checked the last test run and your new test ran successfully.)

**Pull request merged**